src/components/Upload.vue

```diff
@@ -110,17 +121,18 @@ export default {
       const results = [];
       this.uploading = true;
       this.errorMessage = null;
-
+      await this.preUpload();
```

Comment:
How do others feel about reintroducing these hooks? They are duplicative with the design of the Upload class hooks, so they violate the spirit of having only one right way to accomplish something. On the other hand, I see the appeal for these as it makes for less code when you only need an action before the first upload or after the last, which is a common requirement. I am not sure which choice yields greater overall simplicity for users.

Reply:
Allow me to talk in detail my point. I think they are not the same thing, one is for before and after each file and belongs to the uploader and another one is before and after all file upload and belongs to Upload.vue. They are both equally important, there is no straight forward way to achieve one's feature with the other one. 
Let's take [what the demo app is doing with postUpload](https://github.com/girder/girder_web_components/blob/master/demo/App.vue#L144) as an example, the steps I can think of is, first create a Class A that extends the built-in upload.js, then attached the postUpload() to the prototype of the Class A, then bind its context to `this` of App.vue, with arrow function or something. then, debounce the call to the function to only trigger after a certain delay, but it's still not deterministic, since one upload could take a long time, you never know when is the last one. All of these hacking can be subtituded with a simple postUpload() on Upload.vue 
Also, these preUpload() and postUpload() is parallel to the [postUpsert()](https://github.com/girder/girder_web_components/blob/master/demo/App.vue#L151)

Reply:
It's true that this depends on our current status quo where all uploads are done in sequence rather than parallel. I anticipate that we could add a parallel option to the Upload.vue component, which would break this assumption. In light of that, I now agree with you that these are necessary.

Since these will now become the one true way of doing "before all" and "after all" hooks, we should simply remove the `current` and `total` parameters that are passed to the upload class hooks, since we don't see any other utility for them at this time.

Reply:
Sounds good. I'll remove related code.